### PR TITLE
feat(dag-executor): v2 multi-parent registry primitives (Stage 1A)

### PR DESCRIPTION
**Pull request by H31MDALLR**
## Summary

Stage 1A of v2 per-task base chaining, per [I-DAG-MULTI-PARENT-MERGE.md](https://github.com/miniforge-ai/miniforge/blob/main/specs/informative/I-DAG-MULTI-PARENT-MERGE.md) (PR #764, merged).

Pure-data foundation: adds the multi-parent registry helpers and the `:task/merge-strategy` plan-schema slot the v2 orchestrator will consume. **v1 behavior is unchanged on this PR alone** — the new helpers ship but are not yet wired in. Stage 1B (next PR) wires the orchestrator and lights up the new code path.

## What's in this PR

**`branch_registry.clj` Layer 3** (new):
- `multi-parent?` — predicate gating between the single-parent fast path and the multi-parent resolution shape.
- `resolve-multi-parent-base` — returns `{:merge/parents [{:task/id :branch :sha :order}]}` in declaration order. Ancestor collapse stays in the orchestrator (needs git).
- `collapse-duplicate-tips` — drops later parents at the same `:sha`, recording `{:dropped :duplicate-of}` so the orchestrator can log the collapse. Treats `nil :sha` as 'unknown' (never collapses against unknown).
- `compute-input-key` — deterministic SHA-256 hash of `[task-id, strategy, ordered parent SHAs]`, truncated to 16 hex chars. Used to name the namespaced merge ref `refs/miniforge/dag-base/<run-id>/<task-id>/<input-key>`.

**`interface.clj` Layer 9** (new):
- Re-exports: `multi-parent?`, `resolve-multi-parent-base`, `collapse-duplicate-tips`, `compute-merge-input-key`.
- Updated `validate-dag-forest` / `dag-forest?` docstrings: still survive as informational helpers but no longer act as a hard plan-time gate (v2 dropped that gate per spec §10.4).

**`planner.clj`**:
- `PlanTask` schema gains optional `:task/merge-strategy` enum (`:git-merge | :sequential-merge`). Single-parent / zero-dep tasks ignore it.

**Tests**:
- 9 new tests / 22 new assertions in `branch_registry_test.clj` covering ordering, declared-vs-registered skip behavior, duplicate-tip collapse (including nil-SHA edge case), and input-key determinism + sensitivity to each input dimension.

## What's NOT in this PR (deferred to Stage 1B)

- `merge-parent-branches!` — the actual git work (collapse-ancestors, run `git merge`, push to namespaced ref).
- `task-sub-opts` wiring for multi-parent tasks.
- Dropping the v1 forest validator from `execute-plan-as-dag`'s gate.
- Capturing `:sha` at registration time (currently the registry stores branch only).

These all need each other to work end-to-end, so they land together in Stage 1B.

## Test plan

- [x] `bb test` green: 4923 tests / 22621 passes / 0 failures.
- [x] Targeted brick suite green: `clojure -M:test -m cognitect.test-runner -n ai.miniforge.dag-executor.branch-registry-test` → 25 tests / 57 assertions / 0 failures (9 new tests added).
- [ ] Stage 1B will exercise the helpers end-to-end against a real git repo with a 2-parent diamond plan.

## Spec traceability

| Spec section | What this PR delivers |
| ------------ | --------------------- |
| §3.2 step 6 (input-key) | `compute-input-key` |
| §3.2 step 3 (de-duplicate) | `collapse-duplicate-tips` |
| §6.3 duplicate parent tips | `collapse-duplicate-tips` returns `:collapsed` so the orchestrator can emit `:dag/multi-parent-no-op` per spec |
| §10.11 SHA-pinned parents | `resolve-multi-parent-base` returns `:sha` per parent |
| §10.11 strategy rename | `:task/merge-strategy` enum `:git-merge | :sequential-merge` (no `:octopus`) |

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: components/dag-executor/src/ai/miniforge/dag_executor/branch_registry.clj [hunk 1]
```diff
@@ -179,6 +185,122 @@
...
+
+   Returns nil when zero deps would resolve (caller should use the
+   single-parent fast path before calling)."
+  [registry task-deps]
+  (let [resolved (->> task-deps
+                      (map-indexed (fn [order task-id]
+                                     (when-let [info (lookup-branch registry task-id)]
+                                       {:task/id task-id
+                                        :branch  (:branch info)
+                                        :sha     (:sha info)
```
[the remark this answers is not in the document]
Comment (H31MDALLR) [reply components/dag-executor/src/ai/miniforge/dag_executor/branch_registry.clj]: Fixed in 69d792d2 — switched to :commit-sha throughout resolve-multi-parent-base and the rest of Layer 3. Matches the existing convention in register-branch's docstring example, workspace.clj's persist result, and the OCI/docker adapter results. The orchestrator (Stage 1B) will register each completed task with {:branch ... :commit-sha ...} and the multi-parent helpers will see populated SHAs.

File: components/dag-executor/src/ai/miniforge/dag_executor/branch_registry.clj [hunk 2]
```diff
@@ -179,6 +185,122 @@
...
+   the same effective input reuse the same ref instead of accumulating
+   new ones.
+
+   Truncated to 16 hex chars (64 bits) — enough collision resistance
+   for the per-task ref namespace, short enough to read in logs."
+  [task-id strategy parents]
+  (let [canonical (str (pr-str task-id) "|"
+                       (pr-str strategy) "|"
+                       (str/join "," (map :sha parents)))]
+    (subs (hex-digest canonical) 0 16)))
```
[the remark this answers is not in the document]
Comment (H31MDALLR) [reply components/dag-executor/src/ai/miniforge/dag_executor/branch_registry.clj]: Fixed in 69d792d2 — compute-input-key now throws ex-info when any parent has a nil :commit-sha, with :parents-with-nil-sha in ex-data naming the offenders. This is a programming error (the orchestrator must populate SHAs at registration time), not a recoverable anomaly, so failing fast is correct. Each component is also now pr-str'd so the canonical encoding stays unambiguous across all input types. New test compute-input-key-throws-on-nil-commit-sha-test pins the throw + the ex-data shape.

File: components/dag-executor/test/ai/miniforge/dag_executor/branch_registry_test.clj [hunk 3]
```diff
@@ -147,3 +147,118 @@
...
+                  (br/register-branch :a {:branch "task-a" :sha "aaa111"})
+                  (br/register-branch :b {:branch "task-b" :sha "bbb222"})
+                  (br/register-branch :c {:branch "task-c" :sha "ccc333"}))
+          result (br/resolve-multi-parent-base reg [:a :b :c])
+          parents (:merge/parents result)]
+      (is (= 3 (count parents)))
+      (is (= [:a :b :c] (mapv :task/id parents)))
+      (is (= [0 1 2] (mapv :order parents)))
+      (is (= "task-a" (:branch (first parents))))
+      (is (= "aaa111" (:sha (first parents)))))))
```
[the remark this answers is not in the document]
Comment (H31MDALLR) [reply components/dag-executor/test/ai/miniforge/dag_executor/branch_registry_test.clj]: Fixed in 69d792d2 — tests now use :commit-sha throughout to match the existing dag-executor / workspace convention. The renamed test collapse-duplicate-tips-treats-nil-commit-sha-as-unknown-test reflects the key change in its name.

File: components/dag-executor/src/ai/miniforge/dag_executor/branch_registry.clj [hunk 4]
```diff
@@ -35,8 +35,14 @@
 
    Layer 0: Pure registry operations (create / register / lookup)
    Layer 1: Resolution (turn a task's dep set into a base-branch decision)
-   Layer 2: Forest validation (reject multi-parent DAGs at plan time)"
-  (:require [ai.miniforge.messages.interface :as messages]))
+   Layer 2: Forest validation (informational helpers; the v1 plan-time
+            gate is dropped in v2 — multi-parent DAGs are now the
+            normal path, see I-DAG-MULTI-PARENT-MERGE.md)
```
[the remark this answers is not in the document]
Comment (H31MDALLR) [reply components/dag-executor/src/ai/miniforge/dag_executor/branch_registry.clj]: Fixed in 69d792d2 — namespace header now describes Stage 1A behavior accurately. Layer 2 (forest validation) is documented as still-a-gate in 1A, dropped in 1B alongside the orchestrator wiring. Stops misleading readers about what this PR alone changes.

File: components/dag-executor/src/ai/miniforge/dag_executor/interface.clj [hunk 5]
```diff
@@ -571,14 +571,50 @@
...
+   dropped — multi-parent DAGs are the normal path now (see
+   I-DAG-MULTI-PARENT-MERGE.md). `validate-dag-forest` survives as an
+   informational helper for plan-quality reporting."
   branch-registry/validate-forest)
 
 (def dag-forest?
-  "Predicate form of `validate-dag-forest`."
+  "Predicate form of `validate-dag-forest`. Informational only in v2;
+   see [[validate-dag-forest]]."
   branch-registry/forest?)
```
[the remark this answers is not in the document]
Comment (H31MDALLR) [reply components/dag-executor/src/ai/miniforge/dag_executor/interface.clj:591]: Fixed in 69d792d2 — validate-dag-forest and dag-forest? docstrings now describe current Stage 1A behavior. They explicitly state the orchestrator still gates on validate-dag-forest and that the gate is dropped in Stage 1B.
